### ISSUE TITLE
How can I have the evaluation pass in only the response to a prompted/instructed generation into the metric.

### ISSUE THREAD

**Issue by CakeCrusher** [huggingface/trl#1059]
I have created the following metric:
```py
class MyCustomMetric(Metric):
    def _info(self):
        # Returns the MetricInfo that defines the name, description, etc.
        return datasets.MetricInfo(
            # This should be a short description of your metric.
            description="_DESCRIPTION",
            # You can cite papers, GitHub repositories, etc.
            citation="_CITATION",
            # The inputs and outputs your metric expects.
            # These are used to validate the inputs and outputs of _compute
            inputs_description="_KWARGS_DESCRIPTION",
            features=datasets.Features({
                'predictions': datasets.Value('string'),
                'references': datasets.Value('string')
            })
        )

    def _compute(self, predictions, references):
        # Here is where you should put your main metric computation logic
        # Adapt your existing code to fit in here
        
        fc_results = []
        for idx, example in enumerate(predictions):
            print(f"Example {idx}: ", end="")
            post_message = ""

            # Custom Function Calling metric
            prompts = None
            try:
                generated_arguments, expected_arguments, prompts = json_arguments_from_prompt(
                    references[idx],
                    predictions[idx],
                    INSTRUCTION
                    # {"idx": idx, "epoch": epoch}
                )
                fc_result = fc_metric.run(generated_arguments, expected_arguments)

                fc_results.append(fc_result)

                # if save_prompts_path:
                #     # add prompts to dpo_data.json
                #     dpo_data.append({
                #         "fc_result": fc_result,
                #         **prompts
                #     })
                #     with open(save_prompts_path, "w") as f:
                #         json.dump(dpo_data, f)
            except Exception as e:
                print(f"Error function calling: {e}\n")
                fc_results.append(0)
        return fc_results
```
This metric expects the prediction to be generated after passing the instruction. For example I have my prompts in the following format:
`<s> [INST] {message} [/INST] {response}`
I want the evaluation to receive the `predictions` for response and then compare those with my `references`.  To reiterate, the predictions should be generated from the model being passed `<s> [INST] {message} [/INST]`.

Currently it seems as if the logits are just generated without any prompt resulting in responses like:
```
predicted_strings:  ['Unterscheidung Unterscheidung![: What<<NOP What favorite is to help the patterns climate a following is is a to a topic you\nineited by the >>_> in returnFUNCTIONS>\n the is related, return program should be " the format formatname format format. functionFUNCTION_CALL>FORM>( <</OFIGNCIATED_WITH_USER_USERUNCTION</FUNCTION_CALL_NAME>brUNCTIONSCALL_NAMEGSUMENTS>\nGUMENTS_ASS_THE_FIED_FORM_FORMAT</FUNCTION_CALL_ARGUMENTS> If, respond " " response.\nFUNCTIONS>username": "get",meanalth",",function_ "description": "Get health "input": [root": "string", "properties": {" "}] {"name": "leaf_Results", "description": "Search search list of searchists", on a search query", "parameters": {"type": "array", "properties": {"query": {"type": {"query": {"type": "string" "required": "Search"}} "type": "array" "title": ["query"] "description": "Searchphy Search"}}}, {"name": "getUserending",", "description": "Get a list of trifs that on the tr trending", "parameters": {"type": "object", "properties": {"}}},}</FUNCTIONS>\nUSERFS>\n me the ofif from a cat cat doing</users FUNCTION_CALL_NAME>rootSearchResults</FUNCTION_CALL_NAME>FUNCTION_CALL_ARGUMENTS>{"json": {"query": "cool cat"}}</FUNCTION_CALL_ARGUMENTS></s>��']
```

after looking through the source code it seems like modifying the `prediction_step` method inside `Trainer` is the way to go.
